File: sdks/python/examples/basic_usage.py
#!/usr/bin/env python3
"""
Basic usage examples for Task Queue Python SDK
"""

import asyncio
from taskqueue import TaskQueueClient, AsyncTaskQueueClient


def sync_example():
    """Synchronous client example"""
    print("🚀 Synchronous Client Example")
    print("=" * 50)

    # Initialize client
    client = TaskQueueClient(base_url="http://localhost:8080")

    try:
        # Create a task
        task = client.create_task(
            name="Data Processing Example",
            command="python process_data.py --input data.csv --output results.json",
            project_id="550e8400-e29b-41d4-a716-446655440000",  # Example UUID
            description="Process customer data and generate analytics report",
            technical_specs="Python 3.8+, pandas 1.5+, numpy",
            acceptance_criteria=[
                "Process completes without errors",
                "Output file is valid JSON",
                "All data transformations applied correctly"
            ],
            priority="High"
        )

        print(f"✅ Task created: {task.name} (ID: {task.id})")
        print(f"📊 Status: {task.status.value}")
        print(f"⭐ Priority: {task.priority.value}")

        # List tasks
        tasks = client.list_tasks(limit=5)
        print(f"\n📋 Found {len(tasks)} tasks")

        for task in tasks[:3]:  # Show first 3
            print(f"  - {task.name} ({task.status.value})")

    except Exception as e:
        print(f"❌ Error: {e}")


async def async_example():
    """Asynchronous client example"""
    print("\n🚀 Asynchronous Client Example")
    print("=" * 50)

    async with AsyncTaskQueueClient(base_url="http://localhost:8080") as client:
        try:
            # Create multiple tasks
            tasks_data = [
                {
                    "name": "Data Ingestion",
                    "command": "python ingest.py --source api --table users",
                    "project_id": "550e8400-e29b-41d4-a716-446655440000",
                    "description": "Ingest user data from external API",
                    "priority": "Normal"
                },
                {
                    "name": "Data Validation",
                    "command": "python validate.py --input users.json --schema user_schema.json",
                    "project_id": "550e8400-e29b-41d4-a716-446655440000",
                    "description": "Validate ingested data against schema",
                    "priority": "High"
                }
            ]

            created_tasks = await client.create_tasks(tasks_data)

            print(f"✅ Created {len(created_tasks)} tasks:")
            for task in created_tasks:
                print(f"  - {task.name} (ID: {task.id})")

            # Wait for first task to complete (if running)
            if created_tasks:
                first_task = created_tasks[0]
                print(f"\n⏳ Waiting for task '{first_task.name}' to complete...")
                try:
                    completed_task = await client.wait_for_completion(str(first_task.id), timeout=60)
                    print(f"✅ Task completed with status: {completed_task.status.value}")
                except TimeoutError:
                    print("⏰ Task did not complete within timeout")

        except Exception as e:
            print(f"❌ Error: {e}")


async def project_example():
    """Project management example"""
    print("\n🚀 Project Management Example")
    print("=" * 50)

    async with AsyncTaskQueueClient(base_url="http://localhost:8080") as client:
        try:
            # Create a project
            project = await client.create_project(
                name="Customer Analytics Platform",
                description="Platform for analyzing customer behavior and generating insights"
            )

            print(f"✅ Project created: {project.name} (ID: {project.id})")

            # List projects
            projects = await client.list_projects()
            print(f"\n📁 Found {len(projects)} projects:")
            for proj in projects:
                print(f"  - {proj.name} ({proj.status.value})")

        except Exception as e:
            print(f"❌ Error: {e}")


def error_handling_example():
    """Error handling examples"""
    print("\n🚀 Error Handling Example")
    print("=" * 50)

    client = TaskQueueClient(base_url="http://localhost:8080")

    # Example of different error types
    try:
        # This will fail if project doesn't exist
        task = client.create_task(
            name="Test Task",
            command="echo hello",
            project_id="00000000-0000-0000-0000-000000000000"  # Invalid project
        )
    except Exception as e:
        print(f"❌ Caught error: {type(e).__name__}: {e}")

    try:
        # Try to get non-existent task
        task = client.get_task("00000000-0000-0000-0000-000000000000")
    except Exception as e:
        print(f"❌ Caught error: {type(e).__name__}: {e}")


def main():
    """Run all examples"""
    print("Task Queue Python SDK Examples")
    print("=" * 60)

    # Sync examples
    sync_example()

    # Async examples
    asyncio.run(async_example())
    asyncio.run(project_example())

    # Error handling
    error_handling_example()

    print("\n" + "=" * 60)
    print("✅ All examples completed!")


if __name__ == "__main__":
    main()

File: sdks/typescript/examples/basic-usage.ts
// Task Queue TypeScript SDK - Basic Usage Example
import { TaskQueueClient } from '@taskqueue/sdk';

async function main() {
  // Initialize the client
  const client = new TaskQueueClient({
    baseUrl: 'http://localhost:16080',
    apiKey: 'your-api-key', // Optional
  });

  try {
    console.log('🚀 Task Queue SDK Example');
    console.log('========================');

    // Health check
    console.log('\n📊 Checking server health...');
    const health = await client.healthCheck();
    console.log('Health status:', health);

    // Create a project first
    console.log('\n📁 Creating project...');
    const project = await client.createProject({
      name: 'SDK Test Project',
      description: 'Project created by TypeScript SDK',
      tags: ['sdk', 'test'],
    });
    console.log('Created project:', project);

    // Create a task
    console.log('\n📝 Creating task...');
    const task = await client.createTask({
      name: 'Process Data',
      command: 'python process.py --input data.csv --output results.json',
      description: 'Process customer data and generate analytics report',
      project_id: project.id,
      priority: 'High',
      metadata: {
        environment: 'production',
        version: '1.0.0',
      },
    });
    console.log('Created task:', task);

    // List tasks
    console.log('\n📋 Listing tasks...');
    const tasks = await client.listTasks({
      project_id: project.id,
      limit: 10,
    });
    console.log(`Found ${tasks.length} tasks:`);
    tasks.forEach(t => console.log(`  - ${t.name} (${t.status})`));

    // Get task details
    console.log('\n🔍 Getting task details...');
    const taskDetails = await client.getTask(task.id);
    console.log('Task details:', {
      id: taskDetails.id,
      name: taskDetails.name,
      status: taskDetails.status,
      priority: taskDetails.priority,
      created_at: taskDetails.created_at,
    });

    // Update task
    console.log('\n✏️ Updating task...');
    const updatedTask = await client.updateTask(task.id, {
      description: 'Updated description via SDK',
      metadata: {
        ...taskDetails.metadata,
        updated_by: 'sdk',
        updated_at: new Date().toISOString(),
      },
    });
    console.log('Updated task:', updatedTask.name);

    // List projects
    console.log('\n📁 Listing projects...');
    const projects = await client.listProjects();
    console.log(`Found ${projects.length} projects:`);
    projects.forEach(p => console.log(`  - ${p.name} (${p.status})`));

    // Get server metrics
    console.log('\n📊 Getting server metrics...');
    const metrics = await client.getMetrics();
    console.log('Server metrics:', metrics);

    // Cancel task
    console.log('\n❌ Cancelling task...');
    const cancelledTask = await client.cancelTask(task.id, 'Test cancellation via SDK');
    console.log('Cancelled task:', cancelledTask.name);

    console.log('\n✅ SDK example completed successfully!');

  } catch (error) {
    console.error('❌ Error:', error);
    
    if (error instanceof Error) {
      console.error('Error name:', error.name);
      console.error('Error message:', error.message);
    }
  }
}

// Run the example
main().catch(console.error);